huaweicloud/services/servicestage/resource_huaweicloud_servicestage_environment.go

```diff
@@ -0,0 +1,375 @@
+	result := make([]environments.Resource, 0)
+
+	for k, v := range resMap {
+		for _, e := range v.(*schema.Set).List() {
+			result = append(result, environments.Resource{
+				Type: strings.TrimSuffix(k, "_ids"),
+				ID:   e.(string),
+			})
+		}
+	}
+
+	return result
+}
+
+func buildServiceStageEnvCreateOpts(d *schema.ResourceData) environments.CreateOpts {
+	desc := d.Get("description").(string)
+	return environments.CreateOpts{
+		Name:              d.Get("name").(string),
+		Description:       &desc,
+		VpcId:             d.Get("vpc_id").(string),
+		BaseResources:     buildResourcesList(d.Get("basic_resources").([]interface{})),
+		OptionalResources: buildResourcesList(d.Get("optional_resources").([]interface{})),
+	}
+}
+
+func resourceServiceStageEnvCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.ServiceStageV2Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("error creating ServiceStage v2 client: %s", err)
+	}
+
+	opt := buildServiceStageEnvCreateOpts(d)
+	logp.Printf("[DEBUG] The createOpt of ServiceStage environment is: %s", opt)
+	resp, err := environments.Create(client, opt)
+	if err != nil {
+		return fmtp.DiagErrorf("error creating ServiceStage environment: %s", err)
+	}
+
+	d.SetId(resp.ID)
+
+	return resourceServiceStageEnvRead(ctx, d, meta)
+}
+
+func flattenEnvResources(resList []environments.Resource, types []string) []map[string]interface{} {
+	if len(resList) < 1 {
+		return nil
+	}
+
+	result := make(map[string]interface{})
+
+	for _, t := range types {
+		tmp := make([]interface{}, 0)
+		for _, res := range resList {
+			if res.Type == t {
+				tmp = append(tmp, res.ID)
+			}
+		}
+		// Set if and only if the resources set has at least one value to avoid 'terraform apply' triggering an update.
+		if len(tmp) > 0 {
+			result[fmt.Sprintf("%s_ids", t)] = tmp
+		}
+	}
+
+	return []map[string]interface{}{
+		result,
+	}
+}
+
+func resourceServiceStageEnvRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	region := config.GetRegion(d)
+	client, err := config.ServiceStageV2Client(region)
+	if err != nil {
+		return fmtp.DiagErrorf("error creating ServiceStage v2 client: %s", err)
+	}
+
+	resp, err := environments.Get(client, d.Id())
+	if err != nil {
+		common.CheckDeletedDiag(d, err, "error retrieving ServiceStage environment")
+	}
+
+	// Currently, the base resource type supports CCE, CCI, ECS and AS.
+	bTypes := []string{"cce", "cci", "ecs", "as"}
+	// Currently, the optional resource type supports ELB, EIP, RDS, DCS and CSE.
+	oTypes := []string{"elb", "eip", "rds", "dcs", "cse"}
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("name", resp.Name),
+		d.Set("description", resp.Description),
+		d.Set("vpc_id", resp.VpcId),
+		d.Set("basic_resources", flattenEnvResources(resp.BaseResources, bTypes)),
```

Comment:
repeated? basic_resources  and optional_resources

Reply:
yeah, the line 226-227 is useless codes.